lib/SILOptimizer/Utils/ConstExpr.cpp

```diff
@@ -0,0 +1,1258 @@
+
+      // If this is an out-parameter, it is like a store.  If not, this is an
+      // indirect read which is ok.
+      unsigned numIndirectResults = conventions.getNumIndirectSILResults();
+      unsigned opNum = use->getOperandNumber() - 1;
+      if (opNum >= numIndirectResults)
+        continue;
+
+      // Otherwise this is a write. If we have already found a value for this
+      // stack slot then we're done: we don't support multiple assignment.
+      if (memoryObject->getValue().getKind() != SymbolicValue::UninitMemory)
+        return evaluator.getUnknown(addr, UnknownReason::Default);
+
+      // Set `addr` to the address of the memory we want to initialize, so that
+      // the callee can find and initialize it.
+      setValue(addr, memoryAddress);
+
+      // The callee needs to be a direct call to a constant expression.
+      auto callResult = computeCallResult(apply);
+
+      // If the call failed, we're done.
+      if (callResult.hasValue()) {
+        assert(!callResult.getValue().isConstant());
+        memoryObject->setValue(callResult.getValue());
+        return memoryAddress;
+      }
+
+      // computeCallResult will have figured out the result.
+      assert(memoryObject->getValue().isConstant() &&
+             "Should have found a constant result value");
+      continue;
+    }
+
+    // If it is an index_addr, make sure it is a different address from base.
+    if (auto *iai = dyn_cast<IndexAddrInst>(user)) {
+      assert(use->get() == iai->getBase());
+      if (auto *ili = dyn_cast<IntegerLiteralInst>(iai->getIndex())) {
+        if (ili->getValue().getLimitedValue() != 0)
+          continue;
+      }
+      return evaluator.getUnknown(addr, UnknownReason::Default);
+    }
+
+    if (auto *teai = dyn_cast<TupleElementAddrInst>(user)) {
+      // Try finding a writer among the users of `teai`. For example:
+      //   copy_addr %114 to [initialization] %183 : $*Int32
+      //   %179 = alloc_stack $(Int32, Int32, Int32, Int32)
+      //   %183 = tuple_element_addr %179 : $*(Int32, Int32, Int32, Int32), 3
+      //   %191 = tuple_element_addr %179 : $*(Int32, Int32, Int32, Int32), 3
+      //   copy_addr [take] %191 to [initialization] %178 : $*Int32
+      //
+      // The workflow is: when const-evaluating %178, we const-evaluate %191,
+      // which in turn triggers const-evaluating %179, thereby enter this
+      // function, where `addrInst` being %179. Among its users, %191 is not an
+      // initializer, so we skip it (`getSingleWriterAddressValue(teai)` below
+      // will return non-const on it). %183 is a good initializer and can be
+      // const-evaluated (by const-evaluating %114).
+      auto *use = teai->getSingleUse();
+      if (!use)
+        continue;
+
+      auto tupleEltAddrValue = getSingleWriterAddressValue(teai);
+      if (!tupleEltAddrValue.isConstant())
+        continue;
+      // If the tuple elt is indeed a const, we write it into (the appropriate
```

Comment:
@marcrasi  I have a suggestion here. It seems `!tupleEltAddrValue.isConstant()` is used to imply that tupleEltAddrValue has no writers at all is only read. (Like the %191 example presented above.) But, we are sort of overloading the meaning of !isConstnat() here kind of mean read only. I wonder whether this could be make explicit by returning e.g. a boolean flag from the function `getSingleWriterAddressValue()` to indicate that no writer was found. This code can them check for this flag.

Reply:
I was able to write a SIL test that exposes a bug that this causes! Will change it and add the test.